### PR TITLE
bt-172:  Create User Profile Endpoint

### LINKED ISSUE #172: Create UserProfile endpoint
Create endpoint to add user profile information. Flow will be the next - we will save user in every registration step. So finish first step => call PATCH method => finish second step => push PATCH method...
- [x] create user details repository
- [x] create user details service
- [x] create user details controller
- [x] create dto types for endpoint 
- [x] add endpoint (PATCH - method)
- [x] add validation for this method

### PR DESCRIPTION
**Pull request by JosuerBague**
Closes #172

### CODE REVIEW

File: shared/src/bundles/users/validation-schemas/user-details-update.validation-schema.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    EmploymentType,
+    EnglishLevel,
+    JobTitle,
+    NotConsidered,
+    PreferredLanguages,
+} from '../enums/enums.js';
+import { type UserDetailsUpdateRequestDto } from '../types/types.js';
+
+const userDetailsUpdate = joi.object<UserDetailsUpdateRequestDto, true>({
+    id: joi.string().trim().required(),
```
Comment (JosuerBague) [inline shared/src/bundles/users/validation-schemas/user-details-update.validation-schema.ts]: I'm not sure if id should be required. 

Since atm our token only contains user id, which we will use to find this particular user's user details entity.
Also, I don't think there will be a point in time where we will change a user's user details entity's id.

This is my opinion though. What do you think?

Comment (Monyster) [reply shared/src/bundles/users/validation-schemas/user-details-update.validation-schema.ts]: I did as you said. 
But now I think that user can have 2 or more profiles to search for jobs. 
For example, find as a js developer and another as a project manager.
But my opinion differs from reffered platforms.

**Comment by JosuerBague** [comment]
Tested Validation Schema with invalid values, other than experienceYears :
```
{
  "userId": "e7527dde-a9a5-4133-885d-d1d08c5ddc1d",
  "isHired": "1234",
  "profileName": 1234,
  "salaryExpectation": "banana",
  "jobTitle": "Masterchef",
  "location": "Earth",
  "experienceYears": 5,
  "employmentType": {},
  "description": 123,
  "englishLevel": 1,
  "notConsidered": {},
  "preferredLanguages": {}, 
  "projectLinks": 1,
  "photoId": 2,
  "fullName": 3,
  "phone": 4,
  "linkedinLink": 5,
  "companyName": 6,
  "companyLogoId": 7,
  "companyWebsite": 8,
  "employerPosition": 9,
  "cvId": 10
}
```

Validation works.
image

No changes made in database:
image

File: shared/src/enums/error-messages.enum.ts [hunk 2]
```diff
@@ -4,6 +4,9 @@ const ErrorMessages = {
     INCORRECT_EMAIL: 'Incorrect email.',
     EMAIL_ALREADY_EXISTS: 'Email is already taken.',
     PASSWORDS_NOT_MATCH: 'Passwords do not match.',
+    ROUTE_ADMIN_ONLY: 'This route is for administrators only.',
```
Comment (MaksymGavatuk) [inline shared/src/enums/error-messages.enum.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: shared/src/bundles/user-details/validation-schemas/user-details-approve.validation-schema.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+import { type UserDetailsApproveRequestDto } from '../types/types.js';
+
+const userDetailsApprove = joi.object<UserDetailsApproveRequestDto, true>({
+    id: joi.string().trim().required(),
+    isApproved: joi.boolean().required(),
+    deniedReason: joi.string(),
+});
+
+export { userDetailsApprove };
```
Comment (MaksymGavatuk) [inline shared/src/bundles/user-details/validation-schemas/user-details-approve.validation-schema.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: shared/src/bundles/user-details/user-details.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    UserDetailsApiPath,
+} from './enums/enums.js';
+export {
+    type UserDetailsApproveRequestDto,
+    type UserDetailsResponseDto,
+    type UserDetailsUpdateDto,
+    type UserDetailsUpdateRequestDto,
+} from './types/types.js';
+export { userDetailsUpdate as userDetailsUpdateValidationSchema } from './validation-schemas/validation-schemas.js';
+export { userDetailsApprove as userDetailsApproveValidationSchema } from './validation-schemas/validation-schemas.js';
```
Comment (MaksymGavatuk) [inline shared/src/bundles/user-details/user-details.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: shared/src/bundles/user-details/types/user-details-approve-request-dto.type.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+type UserDetailsApproveRequestDto = {
+    id: string;
+    isApproved: boolean;
+    deniedReason?: string;
+};
+
+export { type UserDetailsApproveRequestDto };
```
Comment (MaksymGavatuk) [inline shared/src/bundles/user-details/types/user-details-approve-request-dto.type.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: backend/src/common/types/service.type.ts [hunk 6]
```diff
@@ -4,7 +4,7 @@ type Service<T = unknown> = {
         items: T[];
     }>;
     create(payload: unknown): Promise<T>;
-    update(): Promise<T>;
+    update(payload: Record<string, unknown>): Promise<T>;
```
Comment (MaksymGavatuk) [inline backend/src/common/types/service.type.ts]: Why we can't use type instead of Record<string, unknown> ?

File: backend/src/common/types/repository.type.ts [hunk 7]
```diff
@@ -2,7 +2,7 @@ type Repository<T = unknown> = {
     find(payload: Record<string, unknown>): Promise<T>;
     findAll(): Promise<T[]>;
     create(payload: unknown): Promise<T>;
-    update(): Promise<T>;
+    update(payload: Record<string, unknown>): Promise<T>;
```
Comment (MaksymGavatuk) [inline backend/src/common/types/repository.type.ts]: Why we can't use type instead of Record<string, unknown> ?

File: backend/src/common/middlewares/admin-route.middleware.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+    if (!isAdmin) {
+        throw new HttpError({
+            message: ErrorMessages.ROUTE_ADMIN_ONLY,
+            status: HttpCode.FORBIDDEN,
+        });
+    }
+
+    done();
+};
+
```
Comment (MaksymGavatuk) [inline backend/src/common/middlewares/admin-route.middleware.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: backend/src/bundles/user-details/user-details.model.ts [hunk 9]
```diff
@@ -21,14 +16,19 @@ import {
...
+    type JobTitle,
+    type NotConsidered,
+    type PreferredLanguages,
+} from './enums/enums.js';
 
 class UserDetailsModel extends AbstractModel {
     public 'userId': string;
 
-    public 'isApproved': boolean;
+    public 'isApproved': boolean | null;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.model.ts]: If you made changes in model, you need to make sure we also changed it in db ?

File: backend/src/bundles/auth/auth.controller.ts [hunk 10]
```diff
@@ -115,9 +122,11 @@ class AuthController extends ControllerBase {
             body: UserSignUpRequestDto;
         }>,
     ): Promise<ApiHandlerResponse> {
+        const user = await this.authService.signUp(options.body);
+        await this.userDetailsService.create({ id: user.id });
```
Comment (MaksymGavatuk) [inline backend/src/bundles/auth/auth.controller.ts]: We don't need create userDetails after signUp

File: backend/src/bundles/hard-skills/hard-skills.model.ts [hunk 11]
```diff
@@ -1,12 +1,12 @@
-import { Model } from 'objection';
+// import { Model } from 'objection';
 
-import { UserDetailsModel } from '~/bundles/users/user-details.model.js';
+// import { UserDetailsModel } from '~/bundles/user-details/user-details.js';
```
Comment (MaksymGavatuk) [inline backend/src/bundles/hard-skills/hard-skills.model.ts]: If you don't need this code remove it

File: backend/src/bundles/hard-skills/hard-skills.model.ts [hunk 12]
```diff
@@ -1,12 +1,12 @@
-import { Model } from 'objection';
+// import { Model } from 'objection';
```
Comment (MaksymGavatuk) [inline backend/src/bundles/hard-skills/hard-skills.model.ts]: If you don't need this code remove it

File: backend/src/bundles/hard-skills/hard-skills.model.ts [hunk 13]
```diff
@@ -1,12 +1,12 @@
...
+// import { UserDetailsModel } from '~/bundles/user-details/user-details.js';
 import {
     AbstractModel,
     DatabaseTableName,
-    HardSkillsTableColumn,
-    TalentHardSkillsTableColumn,
-    UserDetailsTableColumn,
+    // HardSkillsTableColumn,
+    // TalentHardSkillsTableColumn,
+    // UserDetailsTableColumn,
```
Comment (MaksymGavatuk) [inline backend/src/bundles/hard-skills/hard-skills.model.ts]: If you don't need this code remove it

File: backend/src/bundles/hard-skills/hard-skills.model.ts [hunk 14]
```diff
@@ -17,18 +17,18 @@ class HardSkillsModel extends AbstractModel {
...
+        //     modelClass: UserDetailsModel,
+        //     join: {
+        //         from: `${DatabaseTableName.HARD_SKILLS}.${HardSkillsTableColumn.ID}`,
+        //         through: {
+        //             from: `${DatabaseTableName.TALENT_HARD_SKILLS}.${TalentHardSkillsTableColumn.HARD_SKILL_ID}`,
+        //             to: `${DatabaseTableName.TALENT_HARD_SKILLS}.${TalentHardSkillsTableColumn.USER_DETAILS_ID}`,
+        //         },
+        //         to: `${DatabaseTableName.USER_DETAILS}.${UserDetailsTableColumn.ID}`,
+        //     },
+        // },
```
Comment (MaksymGavatuk) [inline backend/src/bundles/hard-skills/hard-skills.model.ts]: Why did you remove relation mappings ?

File: backend/src/bundles/user-details/user-details.controller.ts [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+class UserDetailsController extends ControllerBase {
+    private userDetailsService: UserDetailsService;
+
+    public constructor(logger: Logger, userDetailsService: UserDetailsService) {
+        super(logger, ApiPath.USER_DETAILS);
+
+        this.userDetailsService = userDetailsService;
+
+        this.addRoute({
+            path: UserDetailsApiPath.UPDATE,
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.controller.ts]: Path should be root, but method PATCH

File: backend/src/bundles/user-details/user-details.controller.ts [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+            },
+            preHandler: adminRoute,
+            handler: (options) =>
+                this.approve(
+                    options as ApiHandlerOptions<{
+                        body: UserDetailsApproveRequestDto;
+                    }>,
+                ),
+        });
+    }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.controller.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: backend/src/bundles/user-details/user-details.controller.ts [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+        options: ApiHandlerOptions<{
+            body: UserDetailsApproveRequestDto;
+        }>,
+    ): Promise<ApiHandlerResponse> {
+        return {
+            status: HttpCode.OK,
+            payload: await this.userDetailsService.approve(options.body),
+        };
+    }
+}
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.controller.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: backend/src/bundles/user-details/user-details.repository.ts [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+class UserDetailsRepository implements Repository {
+    private userDetailsModel: typeof UserDetailsModel;
+
+    public constructor(userDetailsModel: typeof UserDetailsModel) {
+        this.userDetailsModel = userDetailsModel;
+    }
+
+    public async find(
+        payload: Record<string, unknown>,
+    ): Promise<UserDetailsEntity | undefined> {
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.repository.ts]: instead of undefined return null please
```suggestion
    ): Promise<UserDetailsEntity | null> {
```

File: backend/src/bundles/user-details/user-details.repository.ts [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+        const details = await this.userDetailsModel.query().findOne(payload);
+
+        return details ? UserDetailsEntity.initialize(details) : undefined;
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.repository.ts]:
```suggestion
        return details ? UserDetailsEntity.initialize(details) : null;
```

File: backend/src/bundles/user-details/user-details.service.ts [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+class UserDetailsService implements Service {
+    private userDetailsRepository: UserDetailsRepository;
+
+    public constructor(userDetailsRepository: UserDetailsRepository) {
+        this.userDetailsRepository = userDetailsRepository;
+    }
+
+    public async find(
+        payload: Record<string, unknown>,
+    ): Promise<UserDetailsEntity | undefined> {
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.service.ts]:
```suggestion
    ): Promise<UserDetailsEntity | null> {
```

File: backend/src/bundles/user-details/user-details.service.ts [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+        return this.userDetailsRepository.find({ ...payload });
+    }
+
+    public async findUserDetails(payload: {
+        userId: string;
+    }): Promise<UserDetailsEntity | undefined> {
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.service.ts]:
```suggestion
    }): Promise<UserDetailsEntity | null> {
```

File: backend/src/bundles/user-details/user-details.service.ts [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+            id: userDetailsId,
+            ...rest,
+        });
+    }
+
+    public async approve(
+        payload: UserDetailsApproveRequestDto,
+    ): Promise<UserDetailsEntity | undefined> {
+        return this.userDetailsRepository.update(payload);
+    }
```
Comment (MaksymGavatuk) [inline backend/src/bundles/user-details/user-details.service.ts]: You don't need this in this PR. This is another task. Please don't go beyond the scope of your task !

File: shared/src/bundles/user-details/enums/user-role.enum.ts [hunk 23]
```diff
@@ -0,0 +1,7 @@
+const UserRole = {
+    TALENT: 'talent',
+    EMPLOYER: 'employer',
+    ADMIN: 'admin',
+} as const;
+
+export { UserRole };
```
Comment (MaksymGavatuk) [inline shared/src/bundles/user-details/enums/user-role.enum.ts]: We have such UserRole in users bundles

**Review by MaksymGavatuk — CHANGES_REQUESTED** [review]

File: backend/src/bundles/user-details/user-details.service.ts (hunk 22 above)
Comment (Monyster) [reply backend/src/bundles/user-details/user-details.service.ts]: Can I leave it in this task? Because it has already implemented.

File: backend/src/bundles/user-details/user-details.controller.ts (hunk 17 above)
Comment (Monyster) [reply backend/src/bundles/user-details/user-details.controller.ts]: Can I leave it in this task? Because it has already implemented.

File: backend/src/bundles/user-details/user-details.controller.ts (hunk 16 above)
Comment (Monyster) [reply backend/src/bundles/user-details/user-details.controller.ts]: Can I leave it in this task? Because it has already implemented.

File: backend/src/bundles/hard-skills/hard-skills.model.ts (hunk 14 above)
Comment (Monyster) [reply backend/src/bundles/hard-skills/hard-skills.model.ts]: Backend doesn't work with this code, but I think it's necessary part for the DB logic
image

File: backend/src/bundles/user-details/user-details.controller.ts (hunk 17 above)
Comment (MaksymGavatuk) [reply backend/src/bundles/user-details/user-details.controller.ts]: Separate it to 2 different PR

File: backend/src/bundles/hard-skills/hard-skills.model.ts (hunk 14 above)
Comment (MaksymGavatuk) [reply backend/src/bundles/hard-skills/hard-skills.model.ts]: Got it, try change this logic to: 
modelClass: join(__dirname, pathToModel),

**Comment by Monyster** [comment]
@MaksymGavatuk 
How to build update logic correctly?
I have path to update, but how to update if there are no rows in the userDetails table?
Currently, I return HttpError(Not Found) response.
Do I need to create (create post method)?

**Comment by MaksymGavatuk** [comment]
> @MaksymGavatuk How to build update logic correctly? I have path to update, but how to update if there are no rows in the userDetails table? Currently, I return HttpError(Not Found) response. Do I need to create (create post method)?

Yes, you are right. Add please also create POST endpoint

File: backend/src/bundles/hard-skills/hard-skills.model.ts (hunk 14 above)
Comment (Monyster) [reply backend/src/bundles/hard-skills/hard-skills.model.ts]: Does not help, causes other errors

Comment (Monyster) [reply backend/src/bundles/hard-skills/hard-skills.model.ts]: Temporary fixed. 
```modelClass: "",```

File: backend/src/bundles/hard-skills/hard-skills.model.ts [hunk 24]
```diff
@@ -19,7 +18,7 @@ class HardSkillsModel extends AbstractModel {
     public static override relationMappings = {
         talents: {
             relation: Model.ManyToManyRelation,
-            modelClass: UserDetailsModel,
+            modelClass: '',
```
Comment (nikita-remeslov) [inline backend/src/bundles/hard-skills/hard-skills.model.ts:21]: why do we need these changes?

File: backend/src/bundles/user-details/user-details.repository.ts [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+        payload: UserDetailsFindRequestDto,
+    ): Promise<UserDetailsEntity | null> {
+        const details = await this.userDetailsModel
+            .query()
+            .findOne({ ...payload });
+
+        return details ? UserDetailsEntity.initialize(details) : null;
+    }
+
+    public async findAll(): ReturnType<Repository['findAll']> {
```
Comment (nikita-remeslov) [inline backend/src/bundles/user-details/user-details.repository.ts]: do we need this method?

File: backend/src/bundles/user-details/user-details.repository.ts [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+        const { id, ...rest } = payload;
+
+        const details = await this.userDetailsModel
+            .query()
+            .patchAndFetchById(id, rest);
+
+        return UserDetailsEntity.initialize(details);
+    }
+
+    public async delete(): Promise<boolean> {
```
Comment (nikita-remeslov) [inline backend/src/bundles/user-details/user-details.repository.ts]: do we need this method?

File: backend/src/bundles/user-details/user-details.service.ts [hunk 27]
```diff
@@ -0,0 +1,64 @@
...
+        this.userDetailsRepository = userDetailsRepository;
+    }
+
+    public async find(
+        payload: UserDetailsFindRequestDto,
+    ): Promise<UserDetailsEntity | null> {
+        return this.userDetailsRepository.find({ ...payload });
+    }
+
+    public findAll(): Promise<{ items: unknown[] }> {
```
Comment (nikita-remeslov) [inline backend/src/bundles/user-details/user-details.service.ts:27]: do we need this method?

File: backend/src/bundles/user-details/user-details.service.ts [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+
+        const userDetailsId = userDetails.toObject().id as string;
+
+        return this.userDetailsRepository.update({
+            id: userDetailsId,
+            ...rest,
+        });
+    }
+
+    public delete(): Promise<boolean> {
```
Comment (nikita-remeslov) [inline backend/src/bundles/user-details/user-details.service.ts:59]: do we need this method?

File: shared/src/enums/error-messages.enum.ts [hunk 29]
```diff
@@ -4,6 +4,8 @@ const ErrorMessages = {
     INCORRECT_EMAIL: 'Incorrect email.',
     EMAIL_ALREADY_EXISTS: 'Email is already taken.',
     PASSWORDS_NOT_MATCH: 'Passwords do not match.',
+    NOT_FOUND: 'Not found.',
```
Comment (nikita-remeslov) [inline shared/src/enums/error-messages.enum.ts:7]: Please add dots to all error messages

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/bundles/hard-skills/hard-skills.model.ts (hunk 24 above)
Comment (Monyster) [reply backend/src/bundles/hard-skills/hard-skills.model.ts:21]: Described earlier. Something with path or something else.
https://github.com/BinaryStudioAcademy/bsa-2023-bsa-talents/pull/190#discussion_r1314600310

**Review by nikita-remeslov — APPROVED** [review]

**Review by MaksymGavatuk — APPROVED** [review]
